`Tutorial groups`: Remove custom calendars

az108 opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
The current date and time pickers in the tutorial‐group management pages (creating holidays and manually adding tutorial sessions) use a different UI than the rest of Artemis. This inconsistency is not nice when it comes to usability and breaks the visual flow of the application.

Fixes issue [#10692](https://github.com/ls1intum/Artemis/issues/10692).

### Description
<!-- Describe your changes in detail -->
Replaced those native pickers with our shared jhi‐date‐time‐picker component, thus restoring a uniform look and feel.

### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Instructor
- 1 Course with Tutorial Groups enabled

1.	As Instructor, go to Course Management → Tutorial Groups.
2.	Click on the Holidays tab.
3.	Click Add Holiday.
4.	Verify that the shared datepicker is used and creating a holiday works as expected.
5.	Still as Instructor, go to Tutorial Groups again and click the Sessions tab.
6.	Click Add Session Manually.
7.	Verify that the shared datepicker is used and creating a tutorial group session works as expected.
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
<!-- See [Large Course Setup](https://github.com/ls1intum/Artemis/tree/develop/supporting_scripts/course-scripts/quick-course-setup) for the automation script that handles large course setup. -->
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

old:
<img width="1218" alt="image" src="https://github.com/user-attachments/assets/f213c02b-534a-4966-953a-e0870c3e9212" />
<img width="1142" alt="image" src="https://github.com/user-attachments/assets/7c421eb0-4a93-48d7-a81a-8c5fe3a1b089" />
<img width="1140" alt="image" src="https://github.com/user-attachments/assets/79faa6c0-df96-4169-a6b6-806f57dad20b" />
new:
<img width="732" alt="image" src="https://github.com/user-attachments/assets/457bbfea-4f90-4564-9ef7-b29c837581b9" />
<img width="1439" alt="image" src="https://github.com/user-attachments/assets/5d45853d-1a66-4f50-9bd0-8311db630e8b" />



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Refactor**
  - Simplified and standardized date and time input fields in tutorial group forms by replacing manual inputs with a reusable date-time picker component.
  - Streamlined validation logic for date and time fields, improving consistency and reducing complexity.
  - Updated form submission to explicitly convert date/time values to native Date objects.
  - Refined test setups with synchronous initialization and enhanced translation module integration.
  - Enhanced date-time picker component to properly notify form controls when dates are cleared.

- **Bug Fixes**
  - Improved validation for date and time ranges, ensuring clearer error handling when end dates precede start dates.

- **Style**
  - Enhanced form readability by cleaning up template markup and adjusting indentation.

- **Documentation**
  - Removed obsolete validation messages related to date inputs in English and German localization files.

- **Tests**
  - Simplified tutorial group session and free period form tests by removing outdated helpers and adjusting required field validations.
  - Added new test cases verifying date range validation behavior for tutorial free period forms.
  - Extended edit tutorial group free period tests with cases for initialization and time setting accuracy.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ahmetsenturk reviewed (changes requested):
tested on TS2, I get an error when I click on `Create` button after selecting the date:

<img width="1728" alt="Screenshot 2025-06-04 at 08 32 13" src="https://github.com/user-attachments/assets/9a82afbf-df1c-4132-aa45-99919eb8f732" />

```
sentry.error-handler.ts:53 ERROR TypeError: e.setHours is not a function
    at r.combineDateAndTimeWithAlternativeDate (create-tutorial-grou….component.ts:96:14)
    at r.createTutorialGroupFreePeriod (create-tutorial-grou….component.ts:50:97)
    at create-tutorial-grou…omponent.html:9:116
    at r.submitForm (tutorial-group-free-…component.ts:199:28)
    at tutorial-group-free-…component.html:6:58
handleError	@	sentry.error-handler.ts:53
submitForm	@	tutorial-group-free-…rm.component.ts:199
(anonymous)	@	tutorial-group-free-…rm.component.html:6
```

krusche commented:
Great improvement 👍  why did someone initially use different pickers? 🙈

az108 commented on src/main/webapp/app/tutorialgroup/manage/tutorial-free-periods/crud/tutorial-free-period-form/tutorial-group-free-period-form.component.ts:
```diff
@@ -100,23 +114,21 @@ export class TutorialGroupFreePeriodFormComponent implements OnInit, OnChanges {
+            const endDate = dayjs(this.endDateControl.value).startOf('day');
+            const startDate = dayjs(this.startDateControl.value).startOf('day');
+            return endDate.isAfter(startDate);
         }
```
refactored code

az108 commented on src/main/webapp/app/tutorialgroup/manage/tutorial-group-sessions/crud/tutorial-group-session-form/tutorial-group-session-form.component.ts:
```diff
@@ -103,7 +102,7 @@ export class TutorialGroupSessionFormComponent implements OnInit, OnChanges {
                 startTime: ['13:00:00', [Validators.required]],
                 endTime: ['14:00:00', [Validators.required]],
-                date: [undefined, [Validators.required]],
+                date: [undefined],
```
Because the shared datepicker has its own validator and using both leads to two readlines which looks untidy

az108 replied:
i was referring to the ui, ups

az108 commented:
@krusche Im unsure myself as the shared datepicker is a much better and more responsive choice but I would guess we did not have the shared datepicker at that time. Other than that, I dont see the reason the dev used this approach instead 👀. This component needs to be rewritten in the future tho. The next step would be replacing the timers with the shared timer 😄

AjayvirS reviewed (approved):
Tested on TS1 and the correct datepicker is used

florian-glombik commented on src/main/webapp/app/tutorialgroup/manage/tutorial-free-periods/crud/tutorial-free-period-form/tutorial-group-free-period-form.component.spec.ts:
```diff
@@ -111,13 +100,6 @@ describe('TutorialFreePeriodFormComponent', () => {
         clickSubmit(true);
     }));
 
-    it('should block submit when required property is missing', fakeAsync(() => {
-        const requiredControlNames = ['startDate'];
-        for (const controlName of requiredControlNames) {
-            testFormIsInvalidOnMissingRequiredProperty(controlName);
-        }
-    }));
```
I think that we should keep it

florian-glombik reviewed (changes requested):
I do not understand why we do not enforce in the validators that startDate is required, even if the datepicker enforces that right now, I do not see an issue with enforcing it in the using components, which also means that the implementation does not break if the datepicker implementation is changed

az108 commented on src/main/webapp/app/tutorialgroup/manage/tutorial-free-periods/crud/tutorial-free-period-form/tutorial-group-free-period-form.component.ts:
```diff
@@ -220,7 +231,7 @@ export class TutorialGroupFreePeriodFormComponent implements OnInit, OnChanges {
             return;
         }
         this.form = this.fb.group({
-            startDate: [undefined, [Validators.required]],
+            startDate: [undefined],
```
i dont agree with these validator points as I have implemented it with the shared datepicker validator and implementing it double or custom does not look good in the user interface. The user eventhough this is removed is not able to submit the form till the date is picked

az108 commented on src/main/webapp/app/tutorialgroup/manage/tutorial-group-sessions/crud/tutorial-group-session-form/tutorial-group-session-form.component.spec.ts:
```diff
@@ -106,7 +106,7 @@ describe('TutorialGroupSessionForm', () => {
 
     it('should block submit when required property is missing', fakeAsync(() => {
-        const requiredControlNames = ['startTime', 'endTime', 'date', 'location'];
+        const requiredControlNames = ['startTime', 'endTime', 'location'];
```
this test checks for the validators but as these are integrated via submitForm now, we have different tests covering this case

az108 commented:
@florian-glombik I do not agree with the Validator points. I think you missed something crucial during the review. I have adapted the Validation to be done with the help of the shared datepicker. If we leave the current manual Validators this leads to a big issue in the UI:
![image](https://github.com/user-attachments/assets/9dcc3d73-04c8-4b59-b8ed-43e2fcdd254d)
In my opinion its important that the Form is still working the same way as it did before and I have made sure of that. I removed Validators.required and moved the check. Thus meaning one has to still enter a date before beeing able to submit the Form. One still receives the corresponding error messages if one does not choose a date.

florian-glombik reviewed (changes requested):
I checked it out, and the issue that I expected occurs, the validation for checking whether the form is submitable does not work properly

If you do not want to do this via the validator (which I can understand as the styling looks a bit odd in that case) you can use `isValid` from the `jhi-date-time-picker`

https://github.com/user-attachments/assets/8275fe9a-5099-4667-9ee5-492ef946058d

HawKhiem reviewed (dismissed):
Tested on TS3. Date pickers for holiday and tutorial session works as expected. Also the session created on days marked as holidays are also automatically marked as cancelled.

![Screenshot 2025-06-06 115241](https://github.com/user-attachments/assets/7b812013-8572-45d4-8573-47e272bfc50e)
![Screenshot 2025-06-06 115456](https://github.com/user-attachments/assets/8f9d9e48-0ec1-40c4-a460-d50a0fe41065)

az108 commented:
@florian-glombik Thanks for pointing out the issue. What you mentioned with isValid does not really work as I am already working with the related signals there. I fixed the issue by updating the jhi datepicker which was not fully clearing all values when clearDate was called. This led to the button not greying out in your case. The testcase which I removed tests for Validators. There is no use for that if i remove the validator!

florian-glombik commented:
> @Jan-Thurner This is currently the default for the datepicker. If we want this to be different for the whole application, we should make a follow up PR fixing this. I dont think this should be the scope of this PR even if its a simple change as I dont know how we want to handle it globally.

I'll have an eye on that in case we decide to migrate the component to a primeNg https://primeng.org/ component in the near future 👍

Predixx reviewed (dismissed):
Tested on ts2. Works as expected!

AjayvirS reviewed (approved):
reapprove

florian-glombik reviewed (approved):
Approve code

Thanks for the changes and the fix of the datepicker validation! 👍
